Fixed #35718 -- Add JSONArray to django.db.models.functions.

john-parton opened the pull request:
#### Trac ticket number
Ticket: https://code.djangoproject.com/ticket/35718

#### Branch description
This branch adds a `JSONArray` function so that you can construct json arrays. It works extremely similarly to the existing `JSONObject` function.

#### Checklist
- [x] This PR targets the `main` branch. <!-- Backports will be evaluated and done by mergers, when necessary. -->
- [x] The commit message is written in past tense, mentions the ticket number, and ends with a period.
- [x] I have checked the "Has patch" ticket flag in the Trac system.
- [x] I have added or updated relevant tests.
- [x] I have added or updated relevant docs, including release notes if applicable.
- [ ] I have attached screenshots in both light and dark modes for any UI changes.

john-parton commented:
It turned out to be pretty straightforward to get a basic implementation. I did essentially copy/paste the existing `JSONObject` function to create the new `JSONArray` function. It might be possible to slightly reduce the amount of code duplication with a mixin or some refactoring, but I didn't do that yet, obviously.

john-parton commented:
Added very basic docs, again basically copy/pasted from JSONObject with necessary tweaks to make it make sense. Force pushed update

john-parton commented:
Force-pushed to fix postgres tests failing. Think I got it.

john-parton commented:
Postgres tests still failing. Will try to do more local tests before pushing again.

john-parton commented:
So I figured out the failing test, and it has to do with how the different database engines handle `NULL` (the SQL type) in the function call. The default behavior for sqlite is to replace `NULL` with `null` (the JS type). The default behavior for Oracle and Postgres (with the native json_array) is to _omit_ the NULL values. The older postgres function `jsonb_build_array` replaces `NULL` with `null`.

With both the newer postgres function and oracle, you can actually switch between either behavior by providing `null on null` or `absent on null` clauses.

I went ahead and included `null on null` by default, so the behavior across all the supported databases should be the same/normalized.

Unfortunately, there's not really a way to opt into the `absent on null` behavior.

It might be desirable to have something like

```python
JsonArray(
    ...,
    exclude_nulls=True
)
```

or some other such construction, to use the other behavior. However, we'd have to trigger some kind of warning or wrap the functions somehow with SQLite and older postgres.

john-parton commented:
I don't understand why the tests failed.

I checked out the `main` branch and tried running the test suite with these settings

```python
from test_sqlite import *  # NOQA

DATABASES = {
    "default": {
        "ENGINE": "django.db.backends.postgresql",
        "USER": "user",
        "NAME": "django",
        "PASSWORD": "postgres",
        "HOST": "localhost",
        "PORT": 5436,
        "OPTIONS": {
            "server_side_binding": True,
        },
    },
}
```

Using `./runtests.py --settings test_postgres db_functions.comparison`

I still got errors

```
ERROR: test_textfield (db_functions.comparison.test_json_object.JSONObjectTests.test_textfield)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/john/Code/django/django/db/backends/utils.py", line 105, in _execute
    return self.cursor.execute(sql, params)
    ^^^^^^^^^^^^^^^^^
  File "/home/john/.pyenv/versions/django/lib/python3.12/site-packages/psycopg/cursor.py", line 97, in execute
    raise ex.with_traceback(None)
    ^^^^^^^^^^^^^^^^^
psycopg.errors.IndeterminateDatatype: could not determine data type of parameter $1

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/john/.pyenv/versions/3.12.3/lib/python3.12/unittest/case.py", line 58, in testPartExecutor
    yield
  File "/home/john/.pyenv/versions/3.12.3/lib/python3.12/unittest/case.py", line 634, in run
    self._callTestMethod(testMethod)
    ^^^^^^^^^^^^^^^^^
  File "/home/john/.pyenv/versions/3.12.3/lib/python3.12/unittest/case.py", line 589, in _callTestMethod
    if method() is not None:
    ^^^^^^^^^^^^^^^^^
  File "/home/john/Code/django/tests/db_functions/comparison/test_json_object.py", line 93, in test_textfield
    obj = Article.objects.annotate(json_object=JSONObject(text=F("text"))).first()
    ^^^^^^^^^^^^^^^^^
  File "/home/john/Code/django/django/db/models/query.py", line 1078, in first
    for obj in queryset[:1]:
    ^^^^^^^^^^^^^^^^^
  File "/home/john/Code/django/django/db/models/query.py", line 381, in __iter__
    self._fetch_all()
  File "/home/john/Code/django/django/db/models/query.py", line 1911, in _fetch_all
    self._result_cache = list(self._iterable_class(self))
    ^^^^^^^^^^^^^^^^^
  File "/home/john/Code/django/django/db/models/query.py", line 91, in __iter__
    results = compiler.execute_sql(
    ^^^^^^^^^^^^^^^^^
  File "/home/john/Code/django/django/db/models/sql/compiler.py", line 1585, in execute_sql
    cursor.execute(sql, params)
    ^^^^^^^^^^^^^^^^^
  File "/home/john/Code/django/django/db/backends/utils.py", line 79, in execute
    return self._execute_with_wrappers(
    ^^^^^^^^^^^^^^^^^
  File "/home/john/Code/django/django/db/backends/utils.py", line 92, in _execute_with_wrappers
    return executor(sql, params, many, context)
    ^^^^^^^^^^^^^^^^^
  File "/home/john/Code/django/django/db/backends/utils.py", line 100, in _execute
    with self.db.wrap_database_errors:
    ^^^^^^^^^^^^^^^^^
  File "/home/john/Code/django/django/db/utils.py", line 91, in __exit__
    raise dj_exc_value.with_traceback(traceback) from exc_value
    ^^^^^^^^^^^^^^^^^
  File "/home/john/Code/django/django/db/backends/utils.py", line 105, in _execute
    return self.cursor.execute(sql, params)
    ^^^^^^^^^^^^^^^^^
  File "/home/john/.pyenv/versions/django/lib/python3.12/site-packages/psycopg/cursor.py", line 97, in execute
    raise ex.with_traceback(None)
    ^^^^^^^^^^^^^^^^^
django.db.utils.ProgrammingError: could not determine data type of parameter $1
```

Output of `pip freeze`

```
aiohappyeyeballs==2.4.0
aiohttp==3.10.5
aiosignal==1.3.1
aiosmtpd==1.4.6
alabaster==1.0.0
argon2-cffi==23.1.0
argon2-cffi-bindings==21.2.0
asgiref==3.8.1
atpublic==5.0
attrs==24.2.0
babel==2.16.0
bcrypt==4.2.0
black==24.4.2
certifi==2024.8.30
cffi==1.17.1
charset-normalizer==3.3.2
click==8.1.7
-e git+ssh://git@github.com/john-parton/django.git@aa5293068782dfa2d2173c75c8477f58a9989942#egg=Django
docutils==0.21.2
frozenlist==1.4.1
geoip2==4.8.0
h11==0.14.0
idna==3.8
imagesize==1.4.1
Jinja2==3.1.4
MarkupSafe==2.1.5
maxminddb==2.6.2
multidict==6.0.5
mypy-extensions==1.0.0
numpy==2.1.1
outcome==1.3.0.post0
packaging==24.1
pathspec==0.12.1
pillow==10.4.0
platformdirs==4.2.2
psycopg==3.2.1
pycparser==2.22
Pygments==2.18.0
pylibmc==1.6.3
pymemcache==4.0.0
PySocks==1.7.1
pywatchman==2.0.0
PyYAML==6.0.2
redis==5.0.8
requests==2.32.3
selenium==4.24.0
setuptools==74.1.1
sniffio==1.3.1
snowballstemmer==2.2.0
sortedcontainers==2.4.0
Sphinx==8.0.2
sphinxcontrib-applehelp==2.0.0
sphinxcontrib-devhelp==2.0.0
sphinxcontrib-htmlhelp==2.1.0
sphinxcontrib-jsmath==1.0.1
sphinxcontrib-qthelp==2.0.0
sphinxcontrib-serializinghtml==2.0.0
sqlparse==0.5.1
tblib==3.0.0
trio==0.26.2
trio-websocket==0.11.1
typing_extensions==4.12.2
tzdata==2024.1
urllib3==2.2.2
websocket-client==1.8.0
wsproto==1.2.0
yarl==1.9.9
```

john-parton commented:
Looks like there's an issue I reported here: https://code.djangoproject.com/ticket/35734#comment:4

I am going to hold off on working on this feature until that's resolved.

john-parton commented:
Turns out that psycopg3 is really good at sending the correct type information for server side binding when using the `json_array` function, and really bad when using the `jsonb_build_array`. In order to make sure that the types are properly sent, I just cast all of the variadic arguments. This is only totally necessary if you're using server-side binding, so we could skip that, but I don't particularly like the idea of emitting different SQL based on options in the settings if we can get away with it.

As per my tradition, I included way-to-verbose comments in the code that will need to be pared down.

john-parton commented:
Marked ready for review. It still has the error with Postgres 16+ and server-side bindings, but that's not specific to the code here.

john-parton commented:
Rebased, force-pushed. All checks are passing now (perhaps some issue with `main` or `CI` before.)

john-parton commented on django/db/models/functions/comparison.py:
```diff
@@ -143,6 +143,59 @@ def as_sqlite(self, compiler, connection, **extra_context):
+        return super().as_sql(compiler, connection, **extra_context)
+
+    def as_native(self, compiler, connection, *, returning, **extra_context):
+        # By default, sqlite keeps any NULL values in the array.
+        # PostgresSQL and Oracle remove NULL values from the array by default, but
+        # we can change this behavior by adding the NULL ON NULL clause.
+        # This also matches the postgres behavior of jsonb_build_array.
```
Classic way-too-verbose comment. Let me know how to cut this down.

pauloxnet replied:
IMO you can put the text in a docstring

john-parton replied:
Done

john-parton commented:
Rebase / force-pushed. I believe this is ready for review.

We _might_ want to review https://github.com/django/django/pull/18622 first. It might inform the general approach or "vibes" of this pull request.

john-parton commented:
I reworked the flow-control/logic of the new JsonArray class to match the recent changes to JsonObject (for consistency.)

In particular:

* Inner expressions are explicitly cast to their outputfield on all versions of postgres (similar to how all keys are cast to charfield in the JsonObject class)
* Postgres 16 is checked as true (previously was not features.postgres_16) and newer implementation is returned

Some questions I have:

What if a user has already explicitly casted an element? Does that result in a double cast? That can't be true of the JSONObject class because the keys are passed as `**kwargs`

john-parton commented:
Just updated docs to include a note about SQL `NULL` values mapping to json scalar `null`. This is not the default behavior on many backends, so worth noting for those users familiar with their backends particular behavior.

In the future, it would be nice to add a `on_null` behavior to function.

sarahboyce commented:
> @sarahboyce Could we get someone to look at this? It's starting to get pretty good, I think.

You need to update the flags on the ticket like "Needs tests"

john-parton commented:
Great, thanks.

john-parton commented:
I had accidentally used `ruff` as the linter instead of `black`. Re-ran with `black`, rebased, force-pushed.

john-parton commented:
Ah, my test to check for double-cast doesn't work because I had assumed the cast syntax would always be the postgres-specific one. Egg on face. 

Does anyone have a better suggestion for how to check to make sure that there isn't a cast(cast(...))? Regular expressions suck for this task, but I'm not quite able to wrap my head around the best way to check this.

john-parton commented:
Noticed something interesting.

On the SQLite backend, double-casts end up elided in the final SQL, but on postgres, they're present.

For example

`Cast(Cast("age", output_field=CharField()), output_field=CharField())`

Results in

`("age"::varchar)::varchar` on postgres but 

`CAST("db_functions_author"."age" AS text)` on SQLlite

is this a bug maybe?

I'm wondering what happens if you do something like

Cast(Cast(..., SOME_FIELD), SOME_OTHER_FIELD)


Postgres will pass the object through the SOME_FIELD casting logic, but SQLLite won't and just "skip" it.

john-parton commented:
Updated tests so the check for double-cast only runs only postgres (no other backend needs to wrap items in casts).

Black/isort/push

sarahboyce commented on docs/ref/models/database-functions.txt:
```diff
@@ -163,6 +163,32 @@ and ``comment.modified``.
 
+``JSONArray``
+--------------
+
```
```suggestion

.. versionadded:: 5.2

```
This also needs a 5.2 release note

john-parton replied:
Thanks. None of the other functions in that module had a versionadded, so I just didn't think about it.

sarahboyce replied:
We need a release note for Django 5.2 :+1:

john-parton replied:
I've added a note. Just copy/pasted the release note for JSONObject and tweaked it.

sarahboyce reviewed (commented):
I feel like `jsonb_array_elements` was also discussed in the ticket - was it decided not to add this?

john-parton commented:
> I feel like `jsonb_array_elements` was also discussed in the ticket - was it decided not to add this?

So I left it out for two reasons:

1. I wasn't sure how to implement it on non-postgresql backends.
2. I was trying to keep my PR relatively slim. It turned out that there was a decent amount of subtly to how the json_array implementation works, so I didn't want to go crazy adding every conceivable function.
3. In my code, I actually never made use of jsonb_array_elements. I implemented it for completion, but it turned out to be less useful than json_array.

sarahboyce commented:
We don't have to include it in the PR but it changes this from `Fixed` to `Refs` if it doesn't complete the ticket 👍

laymonage commented on django/db/models/functions/comparison.py:
```diff
@@ -143,6 +143,74 @@ def as_sqlite(self, compiler, connection, **extra_context):
         return super().as_sqlite(compiler, connection, function="MAX", **extra_context)
 
 
+class JSONArray(Func):
+    function = "JSON_ARRAY"
+    output_field = JSONField()
```
Before we add this function, should we assess whether we do want this function in `db.models.functions.comparison`? I know that the existing `JSONObject` function is here, so naturally it does make sense to just put `JSONArray` here as well.

However, given your example use case of `JSONArray`, it seems these functions are more likely to be used for building JSON objects and arrays directly in the database to be returned to the app-level code, and not necessarily to be used for comparison with another database value. (JSON object comparison seems rarely useful 🤔)

Also, consider another upcoming feature #18758, as well as the potential of the other JSON functions you mentioned in the ticket. I think we should create a new module for JSON-related functions, e.g. in `django.db.models.functions.json`, similar to how we have `django.db.models.fields.json` (but they should be exported from the higher-level module, i.e. `django.db.models.functions` as they are now).

If this is agreed, we should move `JSONObject` to that module as well (in a separate PR). It should be safe (I think), since we only document importing from `django.db.models.functions`.

john-parton replied:
I would need some consensus on creating a new module.

It makes sense to me, but I'd like to hear other opinions.

sarahboyce replied:
I like this direction :+1: I think we should do it
@charettes @felixxm what do you think?

felixxm replied:
> @charettes @felixxm what do you think?

Agreed

charettes replied:
Makes sense to me as well.

laymonage replied:
Thanks all! I've opened #18939 for this.

john-parton replied:
I've rebased to move the class to the new module, as well as reorganizing the docs so it's in the "JSON" section with "JSONObject." I tried to keep things alphabetical, more or less.

Haven't made the requested change to the logic yet.

laymonage reviewed (commented):
Thanks for implementing the function (on all databases, wow!) and doing all the research 😄 

For an initial pass I think it looks good, but there's a potential simplification we can do in `as_postgresql`. Also, we might want to have a dedicated `json` module for all the JSON-related functions (see comment for details).

Since the ticket has been updated to only be about adding `JSONArray`, we can probably change the title/commit message back to `Fixed` instead of `Refs`.

I might try to think of other test cases later, but so far, great work @john-parton! 👍

john-parton commented on django/db/models/functions/comparison.py:
```diff
@@ -143,6 +143,74 @@ def as_sqlite(self, compiler, connection, **extra_context):
+        # This is done in all cases for consistency.
+
+        # This logic avoids wrapping any source expression in Cast twice.
+        if all(
+            isinstance(expression, Cast) for expression in self.get_source_expressions()
+        ):
+            casted_obj = self
+        else:
+            casted_obj = self.copy()
+            casted_obj.set_source_expressions(
+                [
+                    (
+                        expression
+                        if isinstance(expression, Cast)
+                        else Cast(expression, expression.output_field)
+                    )
+                    for expression in casted_obj.get_source_expressions()
+                ]
+            )
```
You're right, the conditional is not strictly required.

It's there to avoid the unnecessary copy and set_source_expressions (not just copy)

laymonage replied:
Yeah, but then it means we iterate the source expressions twice in the `else` case: once to build the `isinstance()` for the `all()` check, and once to actually do the casting. (Thrice if you also count Python's iteration in the `all()` function itself.)

Based on the tests, the `else` case seems to be the more common case. Is the tradeoff worth it? 🤔 

I'm not against optimisation – just wanted to make sure it's worth the readability cost, especially when this isn't really a hot code path 😃

john-parton replied:
Still need to simplify this.

john-parton replied:
OK, so I removed the extra check/nesting, and I also made one of the comments a little clearer and put it closer to the actual check in question.

john-parton commented:
Just adding a root comment here as well. I've rebased against main:

* Moved class to new `json` module
* Shuffled documentation around
* Added a release note
* Tweaked the postgres logic to avoid double casting to be a bit simpler, with a slightly more helpful comment

john-parton commented:
Unchecked the `needs documentation` checkbox in Trac.

sarahboyce commented on docs/ref/models/database-functions.txt:
```diff
@@ -841,6 +841,34 @@ that deal with time-parts can be used with ``TimeField``:
 ==============
 
+``JSONArray``
+--------------
```
```suggestion
-------------
```

sarahboyce reviewed (commented):
Thank you for the updates - can you also move the test files to `tests/db_functions/json`? :pray:

john-parton commented:
> Thank you for the updates - can you also move the test files to `tests/db_functions/json`? 🙏

Done. I ran a quick `./runtests.py --settings test_postgres db_functions.json` to confirm everything's still wired up correctly. Rebased/force-pushed.

sarahboyce commented on docs/ref/models/database-functions.txt:
```diff
@@ -841,6 +841,34 @@ that deal with time-parts can be used with ``TimeField``:
+.. class:: JSONArray(*expressions)
+
+Accepts a list of field names or expressions and returns a JSON array
+containing those values. SQL ``NULL`` will be mapped to JSON scalar
+``null`` on all supported backends.
```
```suggestion
containing those values.
```
I am tempted to remove this note as I feel it's more confusing than helpful.
I feel this behavior is [documented here](https://docs.djangoproject.com/en/5.1/topics/db/queries/#storing-and-querying-for-none):
> This only applies to `None` as the top-level value of the field. If `None` is inside a [list](https://docs.python.org/3/library/stdtypes.html#list) or [dict](https://docs.python.org/3/library/stdtypes.html#dict), it will always be interpreted as JSON `null`.

So to me this is expected and we perhaps don't need to call it out?

john-parton replied:
Yeah, I think that's probably the expected behavior given the other supporting documentation and behavior elsewhere.

sarahboyce commented on tests/db_functions/json/test_json_array_object.py:
```diff
@@ -0,0 +1,96 @@
+
+
+@skipUnlessDBFeature("has_json_object_function")
+class JSONArrayObjectTests(TestCase):
```
I think my personal preference is to put these tests in `tests/db_functions/json/test_json_array.py` as we probably won't plan to make a new test file for each combination of database functions

john-parton replied:
I did this, but because JSONObject requires `has_json_object_function`, but Felix's other suggestion said to use `supports_json_field`, I had to include both conditions in the tests that require both JSONObject and JSONArray. Hopefully that makes sense.

sarahboyce reviewed (commented):
Added a couple of small comments before I go away till the new year :christmas_tree:

felixxm commented on django/db/models/functions/json.py:
```diff
@@ -5,6 +5,69 @@
+        return super().as_sql(compiler, connection, **extra_context)
+
+    def as_native(self, compiler, connection, *, returning, **extra_context):
+        """Modify JSON_ARRAY for Oracle and Postgres 16 and later.
+
+        Postgres and Oracle both remove SQL NULL values from the array by
+        default. Adds the NULL ON NULL clause to keep NULL values in the array,
+        mapping them to JSON null values, which matches the behavior of SQLite.
+        """
```
I'd chop the first sentence. Also, `Postgres` -> `PostgreSQL`:
```suggestion
        """
        PostgreSQL 16+ and Oracle remove SQL NULL values from the array by
        default. Adds the NULL ON NULL clause to keep NULL values in the array,
        mapping them to JSON null values, which matches the behavior of SQLite.
        """
```

felixxm commented on tests/db_functions/json/test_json_array.py:
```diff
@@ -0,0 +1,122 @@
+from ..models import Article, Author
+
+
+@skipUnlessDBFeature("has_json_object_function")
```
```suggestion
@skipUnlessDBFeature("supports_json_field")
```

john-parton replied:
is it guaranteed that all backends with `supports_json_field == True` have an implementation for JSONArray? I guess the test suite will tell us

felixxm commented on tests/db_functions/json/test_json_array.py:
```diff
@@ -0,0 +1,122 @@
+
+    @unittest.skipUnless(connection.vendor == "postgresql", "PostgreSQL specific tests")
+    def test_explicit_cast(self):
+        obj = Author.objects.annotate(json_array=JSONArray(Cast("age", CharField())))
+
+        self.assertTrue(
+            bool(re.search(r"::varchar", str(obj.query))),
+        )
+        self.assertFalse(
+            bool(re.search(r"::varchar\)?::varchar", str(obj.query))),
+        )
```
`re.search` seems to be unnecessary. I'd assert actually executed query:
```suggestion
        with self.assertNumQueries(1) as ctx:
            self.assertSequenceEqual(
                Author.objects.annotate(
                    json_array=JSONArray(Cast("age", CharField()))
                ).values("json_array"),
                [   
                    {"json_array": ["30"]},
                    {"json_array": ["30"]},
                ],
            )
        sql = ctx.captured_queries[0]["sql"]
        self.assertIn("::varchar", sql)
        self.assertNotIn("::varchar)::varchar", sql)
```

john-parton replied:
Great! I wasn't familiar with those assertions, nor the captured query pattern, so this is definitely more in line what I would have authored had I known.

felixxm commented on django/db/models/functions/json.py:
```diff
@@ -5,6 +5,69 @@
+    output_field = JSONField()
+
+    def as_sql(self, compiler, connection, **extra_context):
+        if not connection.features.has_json_object_function:
+            raise NotSupportedError(
+                "JSONArray() is not supported on this database backend."
+            )
```
It is `JSONArray` not `JSONObject`. IMO we don't need a separate feature flag for this, we can reuse `supports_json_field`.
```suggestion
        if not connection.features.supports_json_field:
            raise NotSupportedError(
                "JSONFields are not supported on this database backend."
            )
```

john-parton replied:
A lot of boilerplate was copy/pasted from JSONObject(), obviously.

felixxm commented on tests/db_functions/json/test_json_array.py:
```diff
@@ -0,0 +1,122 @@
+        self.assertQuerySetEqual(qs, Author.objects.order_by("-alias"))
+
+
+@skipIfDBFeature("has_json_object_function")
```
```suggestion
@skipIfDBFeature("supports_json_field")
```

felixxm reviewed (commented):
@john-parton Thanks :+1: I left small comments.

john-parton commented:
I applied most of the suggestions using Github, pulled it all down, cleaned up things (mostly black) and made some minor required tweaks.

Rebased, ran tests locally, force pushed.

sarahboyce commented:
buildbot, test on oracle.

sarahboyce commented on django/db/models/functions/json.py:
```diff
@@ -5,6 +5,66 @@
+    function = "JSON_ARRAY"
+    output_field = JSONField()
+
+    def as_sql(self, compiler, connection, **extra_context):
```
```suggestion
    def as_sql(self, compiler, connection, **extra_context):
            if not connection.features.supports_json_field:
                raise NotSupportedError(
                    "JSONFields are not supported on this database backend."
                )
```
Been reviewing https://github.com/django/django/pull/18351/ which has a top level check for json field support - I think this might want the same?
